service/history/workflow/matcher/evaluators.go

```diff
@@ -0,0 +1,275 @@
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/temporalio/sqlparser"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+// Supported Fields
+const (
+	WorkflowID              = searchattribute.WorkflowID
+	WorkflowTypeName        = searchattribute.WorkflowType
+	WorkflowStartTime       = searchattribute.StartTime
+	WorkflowExecutionStatus = searchattribute.ExecutionStatus
+)
+
+type (
+	ExprEvaluator interface {
+		Evaluate(expr sqlparser.Expr) (bool, error)
+	}
+
+	MutableStateMatchEvaluator struct {
+		ms workflow.MutableState
+	}
+)
+
+func newMutableStateMatchEvaluator(ms workflow.MutableState) *MutableStateMatchEvaluator {
+	return &MutableStateMatchEvaluator{ms: ms}
+}
+
+func (m *MutableStateMatchEvaluator) Evaluate(expr sqlparser.Expr) (bool, error) {
+	return m.evaluate(expr)
+}
+
+func (m *MutableStateMatchEvaluator) evaluate(expr sqlparser.Expr) (bool, error) {
+
+	if expr == nil {
+		return false, NewMatcherError("input expression cannot be nil")
+	}
+
+	switch e := (expr).(type) {
+	case *sqlparser.AndExpr:
+		return m.evaluateAnd(e)
+	case *sqlparser.OrExpr:
+		return m.evaluateOr(e)
+	case *sqlparser.ComparisonExpr:
+		return m.evaluateComparison(e)
+	case *sqlparser.RangeCond:
+		return m.evaluateRange(e)
+	case *sqlparser.ParenExpr:
+		return m.evaluate(e.Expr)
+	case *sqlparser.IsExpr:
+		return false, NewMatcherError("%s: 'is' expression", notSupportedErrMessage)
+	case *sqlparser.NotExpr:
+		return false, NewMatcherError("%s: 'not' expression", notSupportedErrMessage)
+	case *sqlparser.FuncExpr:
+		return false, NewMatcherError("%s: function expression", notSupportedErrMessage)
+	case *sqlparser.ColName:
+		return false, NewMatcherError("incomplete expression")
+	default:
+		return false, NewMatcherError("%s: expression of type %T", notSupportedErrMessage, expr)
+	}
+}
+func (m *MutableStateMatchEvaluator) evaluateAnd(expr sqlparser.Expr) (bool, error) {
+	andExpr, ok := expr.(*sqlparser.AndExpr)
+	if !ok {
+		return false, NewMatcherError("%v is not an 'and' expression", sqlparser.String(expr))
+	}
+
+	if leftResult, err := m.evaluate(andExpr.Left); err != nil || !leftResult {
+		return leftResult, err
+	}
+
+	// if left is true, then right must be evaluated
+	return m.evaluate(andExpr.Right)
+}
+
+func (m *MutableStateMatchEvaluator) evaluateOr(expr sqlparser.Expr) (bool, error) {
+	orExpr, ok := expr.(*sqlparser.OrExpr)
+	if !ok {
+		return false, NewMatcherError("%v is not an 'or' expression", sqlparser.String(expr))
+	}
+
+	if leftResult, err := m.evaluate(orExpr.Left); err != nil || leftResult {
+		return leftResult, err
+	}
+	// if left is false, then right must be evaluated
+	return m.evaluate(orExpr.Right)
+}
+
+func (m *MutableStateMatchEvaluator) evaluateComparison(expr sqlparser.Expr) (bool, error) {
+	compExpr, ok := expr.(*sqlparser.ComparisonExpr)
+	if !ok {
+		return false, NewMatcherError("%v is not an 'comparison' expression", sqlparser.String(expr))
+	}
```

Comment:
Better accept the concrete type, you should have casted before entering this function.

```suggestion
func (m *MutableStateMatchEvaluator) evaluateComparison(compExpr *sqlparser.ComparisonExpr) (bool, error) {
```

Reply:
done